File: oj_server/questions/2/go/tail.go


// ======== TAIL ========
func assertEqual(actual, expected string) bool {
    return actual == expected
}

type TestCase struct {
    testFunc func() bool
    inputStr string
}

func saveResult(tests []TestCase, resultPath string) int {
    total, passed := len(tests), 0
    firstFail := ""
    for _, t := range tests {
        if t.testFunc() {
            passed++
        } else if firstFail == "" {
            firstFail = t.inputStr
        }
    }
    f, err := os.Create(resultPath)
    if err != nil {
        fmt.Printf("Error: %v\n", err)
        return 1
    }
    defer f.Close()
    fmt.Fprintf(f, "%d %d %s", passed, total, firstFail)
    if passed == total {
        return 0
    }
    return 1
}

func main() {
    if len(os.Args) < 2 {
        fmt.Println("Usage: program <result_file_path>")
        os.Exit(1)
    }
    path := os.Args[1]
    tests := []TestCase{
        {func() bool { return assertEqual(Solution{}.longestPalindrome("a"), "a") }, `"a"`},
        {func() bool { return assertEqual(Solution{}.longestPalindrome("ab"), "a") }, `"ab"`},
        {func() bool { return assertEqual(Solution{}.longestPalindrome("bb"), "bb") }, `"bb"`},
        {func() bool { return assertEqual(Solution{}.longestPalindrome("aba"), "aba") }, `"aba"`},
        {func() bool { return assertEqual(Solution{}.longestPalindrome("baab"), "baab") }, `"baab"`},
        {func() bool { return assertEqual(Solution{}.longestPalindrome("cbbd"), "bb") }, `"cbbd"`},
        {func() bool { return assertEqual(Solution{}.longestPalindrome("abacdfgdcaba"), "aba") }, `"abacdfgdcaba"`},
        {func() bool { return assertEqual(Solution{}.longestPalindrome("abaxyz"), "aba") }, `"abaxyz"`},
        {func() bool { return assertEqual(Solution{}.longestPalindrome("abaxyzzyxf"), "xyzzyx") }, `"abaxyzzyxf"`},
        {func() bool {
            s := strings.Repeat("a", 500)
            return assertEqual(Solution{}.longestPalindrome(s), s)
        }, `500×"a"`},
    }
    os.Exit(saveResult(tests, path))
}

File: oj_server/questions/2/java/tail.java


// ======== TAIL ========
class Main {
    @FunctionalInterface interface TestFunc { boolean run(); }
    static class TestCase { TestFunc f; String in; TestCase(TestFunc f, String in){this.f=f;this.in=in;} }
    static boolean assertEqual(String a, String e) { return a.equals(e); }
    static int saveResult(List<TestCase> tests, String path) {
        int total = tests.size(), passed = 0; String firstFail = "";
        for (TestCase t : tests) {
            if (t.f.run()) passed++;
            else if (firstFail.isEmpty()) firstFail = t.in;
        }
        try (FileWriter w = new FileWriter(path)) {
            w.write(passed + " " + total + " " + firstFail);
        } catch (IOException ex) {
            return 1;
        }
        return passed == total ? 0 : 1;
    }
    public static void main(String[] args) {
        if (args.length < 1) System.exit(1);
        String p = args[0];
        List<TestCase> tests = Arrays.asList(
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("a"), "a"), "\"a\""),
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("ab"), "a"), "\"ab\""),
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("bb"), "bb"), "\"bb\""),
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("aba"), "aba"), "\"aba\""),
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("baab"), "baab"), "\"baab\""),
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("cbbd"), "bb"), "\"cbbd\""),
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("abacdfgdcaba"), "aba"), "\"abacdfgdcaba\""),
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("abaxyz"), "aba"), "\"abaxyz\""),
            new TestCase(() -> assertEqual(new Solution().longestPalindrome("abaxyzzyxf"), "xyzzyx"), "\"abaxyzzyxf\""),
            new TestCase(() -> {
                char[] arr = new char[500];
                Arrays.fill(arr, 'a');
                String s = new String(arr);
                return assertEqual(new Solution().longestPalindrome(s), s);
            }, "500×\"a\"")
        );
        System.exit(saveResult(tests, p));
    }
}
